zerver/lib/stream_subscription.py

```diff
@@ -313,3 +322,87 @@ def get_subscriptions_for_send_message(
         )
     )
     return query
+
+
+def update_all_subscriber_counts_for_user(
+    user_profile: UserProfile, direction: Literal[1, -1]
+) -> None:
+    """
+    Increment/Decrement number of stream subscribers by 1, when reactivating/deactivating user.
+
+    direction -> 1=increment, -1=decrement
+    """
+    get_user_subscribed_streams(user_profile).update(
+        subscriber_count=F("subscriber_count") + direction
+    )
+
+
+def bulk_update_subscriber_counts(
+    direction: Literal[1, -1],
+    streams: dict[int, set[int]],
+) -> None:
+    """Increment/Decrement number of stream subscribers for multiple users.
+
+    direction -> 1=increment, -1=decrement
+    """
+    if len(streams) == 0:
+        return
+
+    # list of tuples (stream_id, n_subscribers -- to be decremented/incremented)
+    # used as the columns of the temporary table delta_table.
+    stream_delta_values = [
+        (stream_id, len(subscribers) * direction) for stream_id, subscribers in streams.items()
+    ]
+
+    # The goal here is to update subscriber_count in a bulk efficient way,
+    # letting the database handle the deltas to avoid some race conditions.
+    # But unlike update_all_subscriber_counts_for_user which uses F() for a single delta value, we can't
+    # use F() to apply different deltas per row in a single update using ORM,
+    # so we use a raw SQL.
+    query = sql.SQL(
+        """UPDATE {stream_table}
+            SET subscriber_count = {stream_table}.subscriber_count + delta_table.delta
+            FROM (VALUES %s) AS delta_table(id, delta)
+            WHERE {stream_table}.id = delta_table.id;
+        """
```

Comment:
I'm not entirely convinced that this algorithm is safe from races. But it may be good enough in practice. I think we can merge this and then investigate anomalies before actually making use of the value in production; after all, it's currently a field we don't read.